agent/consul/stream/event_publisher.go

```diff
@@ -0,0 +1,330 @@
+	// to be held, EventPublisher.lock MUST always be acquired first.
+	lock sync.RWMutex
+
+	// byToken is an mapping of active Subscriptions indexed by a the token and
+	// a pointer to the request.
+	// When the token is modified all subscriptions under that token will be
+	// reloaded.
+	// A subscription may be unsubscribed by using the pointer to the request.
+	byToken map[string]map[*SubscribeRequest]*Subscription
+}
+
+type changeEvents struct {
+	events []Event
+}
+
+// TopicHandler provides functions which create stream.Events for a topic.
+type TopicHandler struct {
+	// Snapshot creates the necessary events to reproduce the current state and
+	// appends them to the eventBuffer.
+	Snapshot func(*SubscribeRequest, SnapshotAppender) (index uint64, err error)
+	// ProcessChanges accepts a slice of Changes, and builds a slice of events for
+	// those changes.
+	ProcessChanges func(db.ReadTxn, db.Changes) ([]Event, error)
+}
+
+// SnapshotAppender appends groups of events to create a Snapshot of state.
+type SnapshotAppender interface {
+	// Append events to the snapshot.
+	// TODO: document why parameter is a slice instead of a single Event
+	Append(events []Event)
+}
+
+// NewEventPublisher returns an EventPublisher for publishing change events.
+// Handlers are used to convert the memDB changes into events.
+// A goroutine is run in the background to publish events to all subscribes.
+// Cancelling the context will shutdown the goroutine, to free resources,
+// and stop all publishing.
+func NewEventPublisher(ctx context.Context, handlers map[Topic]TopicHandler, snapCacheTTL time.Duration) *EventPublisher {
+	e := &EventPublisher{
+		snapCacheTTL: snapCacheTTL,
+		topicBuffers: make(map[Topic]*eventBuffer),
+		snapCache:    make(map[Topic]map[string]*eventSnapshot),
+		publishCh:    make(chan changeEvents, 64),
+		subscriptions: &subscriptions{
+			byToken: make(map[string]map[*SubscribeRequest]*Subscription),
+		},
+		handlers: handlers,
+	}
+
+	go e.handleUpdates(ctx)
+
+	return e
+}
+
+// PublishChanges to all subscribers. tx is a read-only transaction that captures
+// the state at the time the change happened. The caller must never use the tx once
+// it has been passed to PublishChanged.
+func (e *EventPublisher) PublishChanges(tx db.ReadTxn, changes db.Changes) error {
+	defer tx.Abort()
+
+	var events []Event
+	for topic, handler := range e.handlers {
+		if handler.ProcessChanges != nil {
+			es, err := handler.ProcessChanges(tx, changes)
+			if err != nil {
+				return fmt.Errorf("failed generating events for topic %q: %s", topic, err)
+			}
+			events = append(events, es...)
+		}
+	}
+
+	e.publishCh <- changeEvents{events: events}
+	return nil
+}
+
+func (e *EventPublisher) handleUpdates(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			// TODO: also close all subscriptions so the subscribers are moved
+			// to the new publisher?
+			return
```

Comment:
Calling this out, as it seems important. When the store is replaced any existing subscriptions will be left pointing at the old store and won't receive any new events.

Reply:
Yeah good catch. I don't recall if we handled abandon correctly before but the right thing to do would be to Reset all existing streams so clients abandon their caches and restart from the new snapshot.

This is the only correct thing we can do since the restore might have completely change the state being watched or made the indexes go back in time etc. so all client state also needs to be invalidated and started again.

I recall thinking about this at some point but it's quite possible i didn't actually remember to solve it in my last refactor.

Hmm actually this possible raises an issue with calling `Abandon` on server shutdown above - in the case of a server shutdown, it is sufficient to just terminate the subscriber's connection - their cached state is still valid and they'll be able to continue the subscription without reloading a whole new snapshot on a another server.

By calling abandon at shutdown, we'd be forced to hard-reset all connected streams forcing clients to not only reconnect to a different server, but reconnect and request a whole new snapshot.

In practice I don't think that's _too_ terrible but at scale it could cause rolling restarts to have much higher impact on the other servers unnecessarily.

To fix it to be optimal though we'd need to unplumb Abandon from shutdown and make a different way to signal shutdown and plumb that through both streaming and blocking query loops.

That would be ideal really because IIRC blocking queries will not actually terminate immediately on abandon. they will just wake up and attempt to re-query which may result in them still hanging open after shutdown.

I can understand wanting to break that out into a later PR though - we should just keep a record somehow that we should address it before streaming is "ready".